src/rules/noSelfAssignRule.ts

```diff
@@ -0,0 +1,210 @@
+/**
+ * @license
+ * Copyright 2016 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isArrayLiteralExpression, isBinaryExpression, isElementAccessExpression, isIdentifier,
+    isLiteralExpression, isObjectLiteralExpression, isPropertyAccessExpression, isPropertyDeclaration,
+    isShorthandPropertyAssignment, isSpreadAssignment, isSpreadElement } from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+interface Option {
+    props: boolean;
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-self-assign",
+        description: "This rule is aimed at eliminating self assignments.",
+        optionsDescription: Lint.Utils.dedent`
+            This rule has the option to check properties as well, which is turned on by default.
+            You can turn it off manually.
+            {
+                "no-self-assign": ["error", {"props": false}]
+            }
+        `,
+        rationale: Lint.Utils.dedent`
+            Self assignments have no effect, so probably those are an error due to incomplete refactoring.
+            Those indicate that what you should do is still remaining.
+
+                \`foo = foo;\`
+                \`[bar, baz] = [bar, qiz];\`
+            `,
+        options: {
+            type: "object",
+            properties: {
+                props: {
+                    type: "boolean",
+                },
+            },
+        },
+        optionExamples: [[true, { props: false }]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING(name: string) {
+        return `variable '${name}' is assigned to itself.`;
```

Comment:
This isn't always for variables. Either change it to say the type (variable, property, etc.) or just remove that part of the string.